homeassistant/helpers/floor_registry.py

```diff
@@ -0,0 +1,223 @@
+        old = self.floors[floor_id]
+        changes = {}
+
+        if icon is not UNDEFINED and old.icon != icon:
+            changes["icon"] = icon
+
+        normalized_name = None
+        if name is not UNDEFINED and name != old.name:
+            normalized_name = normalize_floor_name(name)
+            if normalized_name != old.normalized_name and self.async_get_floor_by_name(
+                name
+            ):
+                raise ValueError(
+                    f"The name {name} ({normalized_name}) is already in use"
+                )
+
+            changes["name"] = name
+            changes["normalized_name"] = normalized_name
+
+        if not changes:
+            return old
+
+        new = self.floors[floor_id] = dataclasses.replace(old, **changes)  # type: ignore[arg-type]
+        if normalized_name is not None:
+            self._normalized_name_floor_idx[
+                normalized_name
+            ] = self._normalized_name_floor_idx.pop(old.normalized_name)
+
+        self.async_schedule_save()
+        self.hass.bus.async_fire(
+            EVENT_FLOOR_REGISTRY_UPDATED, {"action": "update", "floor_id": floor_id}
+        )
+
+        return new
+
+    async def async_load(self) -> None:
+        """Load the floor registry."""
+        data = await self._store.async_load()
+        floors: MutableMapping[str, FloorEntry] = {}
+
+        if data is not None:
+            for floor in data["floors"]:
+                normalized_name = normalize_floor_name(floor["name"])
+                floors[floor["floor_id"]] = FloorEntry(
+                    icon=floor["icon"],
+                    floor_id=floor["floor_id"],
+                    name=floor["name"],
+                    normalized_name=normalized_name,
+                )
+                self._normalized_name_floor_idx[normalized_name] = floor["floor_id"]
+
+        self.floors = floors
+
+    @callback
+    def async_schedule_save(self) -> None:
+        """Schedule saving the floor registry."""
+        self._store.async_delay_save(self._data_to_save, SAVE_DELAY)
+
+    @callback
+    def _data_to_save(self) -> dict[str, list[dict[str, str | None]]]:
+        """Return data of floor registry to store in a file."""
+        return {
+            "floors": [
+                {
+                    "icon": entry.icon,
+                    "floor_id": entry.floor_id,
+                    "name": entry.name,
+                }
+                for entry in self.floors.values()
+            ]
+        }
+
+
+@callback
+def async_get(hass: HomeAssistant) -> FloorRegistry:
+    """Get floor registry."""
+    return cast(FloorRegistry, hass.data[DATA_REGISTRY])
+
+
+async def async_load(hass: HomeAssistant) -> None:
+    """Load floor registry."""
+    assert DATA_REGISTRY not in hass.data
+    hass.data[DATA_REGISTRY] = FloorRegistry(hass)
+    await hass.data[DATA_REGISTRY].async_load()
+
+
+def normalize_floor_name(floor_name: str) -> str:
```

Comment:
Make this a protected `staticmethod` of `FloorRegistry`

Reply:
With the latest change, I've not moved it. I found it odd to use it in both FloorRegistry & FloorRegistryItems. Instead, I've left it where it is and marked it protected.